File: LondonFhirService.Core/Services/Foundations/FhirRecordDifferences/FhirRecordDifferenceService.cs
// ---------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using LondonFhirService.Core.Brokers.DateTimes;
using LondonFhirService.Core.Brokers.Loggings;
using LondonFhirService.Core.Brokers.Securities;
using LondonFhirService.Core.Brokers.Storages.Sql;
using LondonFhirService.Core.Models.Foundations.FhirRecordDifferences;

namespace LondonFhirService.Core.Services.Foundations.FhirRecordDifferences
{
    internal partial class FhirRecordDifferenceService : IFhirRecordDifferenceService
    {
        private readonly IStorageBroker storageBroker;
        private readonly IDateTimeBroker dateTimeBroker;
        private readonly ISecurityAuditBroker securityAuditBroker;
        private readonly ILoggingBroker loggingBroker;

        public FhirRecordDifferenceService(
            IStorageBroker storageBroker,
            IDateTimeBroker dateTimeBroker,
            ISecurityAuditBroker securityAuditBroker,
            ILoggingBroker loggingBroker)
        {
            this.storageBroker = storageBroker;
            this.dateTimeBroker = dateTimeBroker;
            this.securityAuditBroker = securityAuditBroker;
            this.loggingBroker = loggingBroker;
        }

        public ValueTask<FhirRecordDifference> AddFhirRecordDifferenceAsync(
            FhirRecordDifference fhirRecordDifference,
            CancellationToken cancellationToken = default) =>
            TryCatch(async () =>
            {
                fhirRecordDifference = await this.securityAuditBroker.ApplyAddAuditValuesAsync(fhirRecordDifference);
                await ValidateFhirRecordDifferenceOnAdd(fhirRecordDifference);

                return await this.storageBroker.InsertFhirRecordDifferenceAsync(fhirRecordDifference, cancellationToken);
            });

        public ValueTask<IQueryable<FhirRecordDifference>> RetrieveAllFhirRecordDifferencesAsync(
            CancellationToken cancellationToken = default) =>
            TryCatch(async () => await this.storageBroker.SelectAllFhirRecordDifferencesAsync(cancellationToken));

        public ValueTask<FhirRecordDifference> RetrieveFhirRecordDifferenceByIdAsync(
            Guid fhirRecordDifferenceId,
            CancellationToken cancellationToken = default) =>
            TryCatch(async () =>
            {
                ValidateFhirRecordDifferenceId(fhirRecordDifferenceId);

                FhirRecordDifference maybeFhirRecordDifference = await this.storageBroker
                    .SelectFhirRecordDifferenceByIdAsync(fhirRecordDifferenceId, cancellationToken);

                ValidateStorageFhirRecordDifference(maybeFhirRecordDifference, fhirRecordDifferenceId);

                return maybeFhirRecordDifference;
            });

        public ValueTask<FhirRecordDifference> ModifyFhirRecordDifferenceAsync(
            FhirRecordDifference fhirRecordDifference,
            CancellationToken cancellationToken = default) =>
            TryCatch(async () =>
            {
                fhirRecordDifference = await this.securityAuditBroker.ApplyModifyAuditValuesAsync(fhirRecordDifference);

                await ValidateFhirRecordDifferenceOnModify(fhirRecordDifference);

                FhirRecordDifference maybeFhirRecordDifference =
                    await this.storageBroker.SelectFhirRecordDifferenceByIdAsync(fhirRecordDifference.Id, cancellationToken);

                ValidateStorageFhirRecordDifference(maybeFhirRecordDifference, fhirRecordDifference.Id);

                fhirRecordDifference = await this.securityAuditBroker
                    .EnsureAddAuditValuesRemainsUnchangedOnModifyAsync(fhirRecordDifference, maybeFhirRecordDifference);

                ValidateAgainstStorageFhirRecordDifferenceOnModify(
                    inputFhirRecordDifference: fhirRecordDifference,
                    storageFhirRecordDifference: maybeFhirRecordDifference);

                return await this.storageBroker.UpdateFhirRecordDifferenceAsync(fhirRecordDifference, cancellationToken);
            });

        public ValueTask<FhirRecordDifference> RemoveFhirRecordDifferenceByIdAsync(
            Guid fhirRecordDifferenceId,
            CancellationToken cancellationToken = default) =>
            TryCatch(async () =>
            {
                ValidateFhirRecordDifferenceId(fhirRecordDifferenceId);

                FhirRecordDifference maybeFhirRecordDifference = await this.storageBroker
                    .SelectFhirRecordDifferenceByIdAsync(fhirRecordDifferenceId, cancellationToken);

                ValidateStorageFhirRecordDifference(maybeFhirRecordDifference, fhirRecordDifferenceId);

                return await this.storageBroker.DeleteFhirRecordDifferenceAsync(maybeFhirRecordDifference, cancellationToken);
            });
    }
}

File: LondonFhirService.Manage.Client/src/services/foundations/fhirRecordDifferences/fhirRecordDifferenceService.ts
import { FhirRecordDifferenceApiBroker } from "../../../brokers/apis/fhirRecordDifferenceApiBroker";
import { tryCatchFhirRecordDifferenceServiceAsync } from "./fhirRecordDifferenceService.exceptions";
import {
    validateFhirRecordDifferenceId,
    validateFhirRecordDifferenceModification,
    validateFhirRecordDifferenceQuery
} from "./fhirRecordDifferenceService.validations";
import type { FhirRecordDifference } from "../../../models/foundations/fhirRecordDifferences/FhirRecordDifference";
import type { FhirRecordDifferenceQuery } from "../../../models/foundations/fhirRecordDifferences/FhirRecordDifferenceQuery";
import type { IFhirRecordDifferenceApiBroker } from "../../../brokers/apis/iFhirRecordDifferenceApiBroker";
import type { IFhirRecordDifferenceService } from "./iFhirRecordDifferenceService";

export class FhirRecordDifferenceService implements IFhirRecordDifferenceService {
    private readonly fhirRecordDifferenceApiBroker: IFhirRecordDifferenceApiBroker;

    constructor(
        fhirRecordDifferenceApiBroker: IFhirRecordDifferenceApiBroker =
        new FhirRecordDifferenceApiBroker()) {
        this.fhirRecordDifferenceApiBroker = fhirRecordDifferenceApiBroker;
    }

    public async retrieveFhirRecordDifferencesAsync(
        fhirRecordDifferenceQuery: FhirRecordDifferenceQuery,
        abortSignal?: AbortSignal)
        : Promise<FhirRecordDifference[]> {
        return await tryCatchFhirRecordDifferenceServiceAsync(async () => {
            validateFhirRecordDifferenceQuery(fhirRecordDifferenceQuery);

            return await this.fhirRecordDifferenceApiBroker
                .getFhirRecordDifferencesAsync(fhirRecordDifferenceQuery, abortSignal);
        });
    }

    public async retrieveFhirRecordDifferenceByIdAsync(
        fhirRecordDifferenceId: string,
        abortSignal?: AbortSignal)
        : Promise<FhirRecordDifference> {
        return await tryCatchFhirRecordDifferenceServiceAsync(async () => {
            validateFhirRecordDifferenceId(fhirRecordDifferenceId);

            return await this.fhirRecordDifferenceApiBroker
                .getFhirRecordDifferenceByIdAsync(fhirRecordDifferenceId, abortSignal);
        });
    }

    public async modifyFhirRecordDifferenceAsync(
        fhirRecordDifference: FhirRecordDifference)
        : Promise<FhirRecordDifference> {
        return await tryCatchFhirRecordDifferenceServiceAsync(async () => {
            validateFhirRecordDifferenceModification(fhirRecordDifference);

            return await this.fhirRecordDifferenceApiBroker
                .putFhirRecordDifferenceAsync(fhirRecordDifference);
        });
    }
}
